### PR TITLE
Optimize CPU RAM peak memory during quantization

### PR DESCRIPTION
**Pull request by lvliang-intel**
### Optimize CPU RAM peak memory during quantization:
1. Two optional CPU RAM optimizations, gated by [low_cpu_mem_usage]: 
**cpu_stream_offload_blocks**: offload block weights to disk and load them on demand during block-wise quantization, then re-offload quantized weights; restore at the end.
**cpu_stream_loss**: avoid caching block outputs by computing targets on-the-fly with a frozen block copy (requires [nblocks=1]).

2. The quantization flow caches inputs once, then processes blocks sequentially, loading/offloading weights and optionally streaming loss to keep peak CPU RAM low.

### Test
Quantize Qwen/Qwen3-4B-Instruct-2507 with AutoRound (4-bit) and compare CPU RAM peak usage with different optimization options.

Optimization options:
1. cpu_stream_offload_blocks: Offload block weights to disk, load on demand
2. cpu_stream_loss: Compute loss on-the-fly using frozen block copy

**Summary: Peak RAM Comparison**

| Configuration        | Peak RAM (GB) | Time (s) | RAM Saved |
|---------------------|------------------|----------|-----------|
| Baseline                | 24.29                 | 1582.3   | baseline  |
| + offload_blocks  | 20.26                 | 1609.1   | -4.03 GB  |
| + stream_loss       | 21.31                | 1364.0   | -2.98 GB  |
| All optimizations  | 15.57                | 1269.3   | -8.72 GB  |

## Type of Change

- [ ] Bug fix
- [ ] New feature
- [ ] Documentation update
- [x] Performance improvement
- [ ] Code refactoring
- [ ] Other (please specify):

## Related Issues

[](https://github.com/intel/auto-round/issues/912)

Fixes or relates to #

## Checklist Before Submitting

- [x] My code has been tested locally.
- [ ] Documentation has been updated as needed.
- [x] New or updated tests are included where applicable.

### CODE REVIEW

File: auto_round/compressors/base.py [hunk 1]
```diff
@@ -1517,6 +1528,196 @@ def _quantize_via_rtn_blockwise(self, all_to_quantized_module_names: list[str])
         # if self.is_immediate_saving:
         #     shard_writer(self, is_finalize=True)
 
+    def _estimate_tensor_size_gb(self, tensor) -> float:
```
Comment (n1ck-guo) [inline auto_round/compressors/base.py]: I prefer to move these functions to utils/device.py or utils/model.py

**Comment by lvliang-intel** [comment]
Support AutoScheme CPU RAM Optimization:

CUDA_VISIBLE_DEVICES=0 python compare_auto_scheme_ram.py     --model /models/Qwen2.5-3B-Instruct/     --nsamples 8 --seqlen 256 --batch-size 1
[Result] low_cpu_mem_usage=False -> 'peak_ram': 13.65GB, 'peak_vram': 3.44GB
[Result] low_cpu_mem_usage=False -> time=688.6s
[Result] low_cpu_mem_usage=True -> 'peak_ram': 8.83GB, 'peak_vram': 3.44GB
[Result] low_cpu_mem_usage=True -> time=637.2s
=== Summary ===
Configuration         Peak RAM (GB)   Time (s)    RAM Saved
Baseline                      13.65              688.6               --
Optimizations             8.83                637.2           4.82 GB
Ratio                           0.65x

File: auto_round/auto_scheme/delta_loss.py [hunk 2]
```diff
@@ -55,11 +60,264 @@
...
+    This allows block weights to stay on disk between scheme iterations,
+    keeping only one block in RAM at a time.
+    """
+
+    def __init__(self, low_cpu_mem_usage: bool = False):
+        self.low_cpu_mem_usage = low_cpu_mem_usage
+        # Wrapped-state storage (changes every scheme iteration)
+        self._offload_tempdir: Optional[str] = None
+        self._offloaded_blocks: dict[str, dict] = {}
+        # Original-state storage (saved once, read many)
```
Comment (wenhuach21) [inline auto_round/auto_scheme/delta_loss.py]: is it possible to encapsulate all functionalities into one or more classes and move them to one file, so that modifications in the compressor and delta_loss are as simple as possible?

**Comment by wenhuach21** [comment]
I don't u

> Support AutoScheme CPU RAM Optimization:
> 
> CUDA_VISIBLE_DEVICES=0 python compare_auto_scheme_ram.py --model /models/Qwen2.5-3B-Instruct/ --nsamples 8 --seqlen 256 --batch-size 1 [Result] low_cpu_mem_usage=False -> 'peak_ram': 13.65GB, 'peak_vram': 3.44GB [Result] low_cpu_mem_usage=False -> time=688.6s [Result] low_cpu_mem_usage=True -> 'peak_ram': 8.83GB, 'peak_vram': 3.44GB [Result] low_cpu_mem_usage=True -> time=637.2s === Summary ===
> 
> Configuration	Peak RAM (GB)	Time (s)	RAM Saved
> Baseline	13.65	688.6	--
> Optimizations	8.83	637.2	4.82 GB
> **Ratio**	**0.65x**	**0.92x**

I don't understand why time cost is reduced as well?  Based on my understanding, it should be increased

File: auto_round/auto_scheme/delta_loss.py [hunk 3]
```diff
@@ -55,11 +60,264 @@
...
+
+    # ------------------------------------------------------------------
+    # Directory helpers
+    # ------------------------------------------------------------------
+    def init_offload_dir(self) -> Optional[str]:
+        """Initialize the temporary directory for wrapped-state offload."""
+        if not self.low_cpu_mem_usage:
+            return None
+        if self._offload_tempdir is None:
+            self._offload_tempdir = tempfile.mkdtemp(prefix="autoscheme_offload_")
```
Comment (wenhuach21) [inline auto_round/auto_scheme/delta_loss.py]: I don't think offload is needed for delta_loss, as we have not changed the weights, what we only need to do is releasing the blocks weights, and reload it from model folder again

File: auto_round/auto_scheme/delta_loss.py [hunk 4]
```diff
@@ -817,10 +991,17 @@ def check_bf16_scheme(scheme):
         options_scores.append(options_total_loss)
         clear_memory(device_list=device_list)
 
+    # Restore original weights from disk for final bit-budget computations
+    if offload_context is not None:
+        offload_context.load_all_original_blocks(model, block_name)
```
Comment (wenhuach21) [inline auto_round/auto_scheme/delta_loss.py]: the code change is still complex.
is it possible to register some pre-forward/forward hook to tirgger the reload and release, thus I guess the code change will be very simple

File: auto_round/auto_scheme/gen_auto_scheme.py [hunk 5]
```diff
@@ -40,6 +40,7 @@ class AutoScheme:
     enable_torch_compile: Optional[bool] = None
     disable_opt_rtn: bool = True
     low_gpu_mem_usage: bool = True
+    low_cpu_mem_usage: bool = False  # Enable disk offload for CPU RAM optimization
```
Comment (wenhuach21) [inline auto_round/auto_scheme/gen_auto_scheme.py]: better set it to Ture in auto-scheme

File: auto_round/compressors/base.py [hunk 6]
```diff
@@ -1414,6 +1450,27 @@ def _quantize_rtn(self) -> tuple[torch.nn.Module, dict[str, Any]]:
...
         return self.model, self.layer_config
 
+    def _stream_offload_blocks(self, all_blocks: list[list[str]]) -> None:
+        """Offload all block weights to disk upfront to minimize peak CPU RAM.
+
+        Each block's full ``state_dict`` is saved and then cleared from memory.
+        """
+        if not (self.low_cpu_mem_usage and self.cpu_stream_offload_blocks):
+            return
+        logger.info("stream offloading block weights to disk...")
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: better move this function to offload class

Comment (lvliang-intel) [reply auto_round/compressors/base.py]: ok, let me refine it.

File: auto_round/auto_scheme/gen_auto_scheme.py (hunk 5 above)
Comment (lvliang-intel) [reply auto_round/auto_scheme/gen_auto_scheme.py]: ok

File: auto_round/compressors/base.py [hunk 7]
```diff
@@ -3072,6 +3219,16 @@ def _quantize_block(
 
             return q_outputs, output
         else:
+            # When stream_loss is enabled, output is None, so we need to compute it for the next block
+            if stream_loss and output is None:
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: We should rethink supporting stream loss if it complicates the codebase. The performance gains are negligible for large-scale models and MoE architectures, making the added complexity hard to justify

File: auto_round/utils/offload.py [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+    # Skip WrapperLayer – its weight is a property delegating to orig_layer.
+    if hasattr(module, "orig_layer"):
+        return
+
+    with torch.no_grad():
+        if hasattr(module, "weight") and module.weight is not None:
+            if cache_numel and module.weight.numel() > 0:
+                module._cached_weight_numel = module.weight.numel()
+                module._cached_weight_shape = tuple(module.weight.shape)
+            if isinstance(module.weight, torch.nn.Parameter):
```
Comment (wenhuach21) [inline auto_round/utils/offload.py]: How about using named_parameters()? It would make the code much more robust since the current code only considered weights and bias

Comment (wenhuach21) [reply auto_round/utils/offload.py]: maybe named_buffers are also needed to be taken into account

Comment (lvliang-intel) [reply auto_round/utils/offload.py]: ok

File: auto_round/compressors/base.py [hunk 9]
```diff
@@ -1735,18 +1775,28 @@ def _should_disable_inplace_due_to_layers_outside_block() -> bool:
...
         self.model = mv_module_from_gpu(self.model)
         clear_memory(device_list=self.device_list)
-        if hasattr(self.model, "hf_device_map") and len(self.model.hf_device_map) > 1:
-            accelerate.hooks.remove_hook_from_submodules(self.model)  # self.model.hf_device_map has not been changed
         logger.info("caching done")
+        # Log memory breakdown for model weights
+        if self.low_cpu_mem_usage:
+            model_size_gb = BlockOffloadManager.estimate_model_size_gb(self.model)
+        if self.low_cpu_mem_usage:
+            self._offloader.stream_offload_all_blocks(self.model, all_blocks, self.device_list)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: could we refine the API name? we only need two main APIs, one is offload, the other one is reload (you could change the names). Both of them could take module or list of modules as inputs

File: auto_round/compressors/base.py [hunk 10]
```diff
@@ -1792,6 +1842,10 @@ def _should_disable_inplace_due_to_layers_outside_block() -> bool:
         if self.is_immediate_saving:
             shard_writer(self, is_finalize=True)
 
+        if self.low_cpu_mem_usage:
+            self._offloader.restore_all(self.model)
+            self._offloader.cleanup()
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: could we  track the state and trigger the cleanup automatically thus easily other developer to use such API.

File: auto_round/compressors/base.py [hunk 11]
```diff
@@ -2864,6 +2918,11 @@ def _quantize_block(
                 block, input_ids, input_others, self.batch_size * self.infer_bs_coeff, device, self.cache_device
             )
 
+            # Log output cache size for first block
+            if self.low_cpu_mem_usage and not hasattr(self, "_logged_output_size"):
+                output_size = BlockOffloadManager.estimate_tensor_size_gb(output)
+                self._logged_output_size = True
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: If it is meant to check whether there is enough disk space, then keeping it makes sense. However, if that is not its purpose, we should reconsider it. Since the logic in this function is quite complicated, it would be better to remove it unless it is truly necessary.

File: auto_round/utils/offload.py [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+        self._remove_hooks()
+        self._cleared_blocks.clear()
+
+
+# =====================================================================
+# BlockOffloadManager
+# =====================================================================
+
+
+class BlockOffloadManager:
```
Comment (wenhuach21) [inline auto_round/utils/offload.py]: Is it possible to merge these two classes? We could introduce an argument to control the offload mode (e.g., offload vs. clean) and handle the orig_layer issue through name conversion.

File: auto_round/utils/offload.py [hunk 13]
```diff
@@ -12,15 +12,34 @@
...
+and **reload** it on demand.  Two operating modes are supported:
 
-It also provides ``load_block_from_model_files`` for reloading block weights
-directly from the original HuggingFace model checkpoint (safetensors or
-pytorch bin files) — eliminating the need to write an extra copy to a temp
-directory when the original weights have not been modified.
+* ``"offload"`` -- saves the module's ``state_dict`` to a temporary directory
+  before clearing. Best when the original weights may have been modified
+  (e.g. after quantization) and need to be restored later.
+* ``"clean"``   -- simply clears weights without saving; reloads from the
```
Comment (wenhuach21) [inline auto_round/utils/offload.py:24]: how about offload->_offload, and rename offload_all to offload(names=list|str|list[list])

File: auto_round/utils/offload.py [hunk 14]
```diff
@@ -187,489 +165,534 @@ def _build_weight_map(model_dir: str) -> dict[str, str]:
...
+            self._load_from_disk(name, module)
+            self._remove_saved_entry(name)
+        else:
+            if self.model_dir is None:
+                logger.warning("OffloadManager: model_dir is required for clean mode")
+                return
+            load_block_from_model_files(self.model_dir, name, module)
 
+    def resave(self, model: torch.nn.Module, name: str) -> None:
+        """Discard old save, re-save the (possibly modified) module, and clear it.
```
Comment (wenhuach21) [inline auto_round/utils/offload.py]: could this function be merged to offload, with an extra argument, overwrite=True?

File: auto_round/utils/offload.py [hunk 15]
```diff
@@ -12,15 +12,34 @@
...
+
+    offloader = OffloadManager(mode="offload")
+    offloader.offload(model, "model.layers.0")   # save + clear
+    offloader.reload(model, "model.layers.0")    # load back + auto-clean temp file
+
+**AutoScheme (clean mode with hooks)**::
+
+    offloader = OffloadManager(mode="clean",
+                         model_dir="/path/to/model")
+    offloader.add_hooks(model, block_names)       # clear all + register hooks
```
Comment (wenhuach21) [inline auto_round/utils/offload.py]: is it clearer to rename as add_offload_hooks

File: auto_round/utils/offload.py [hunk 16]
```diff
@@ -187,489 +165,534 @@ def _build_weight_map(model_dir: str) -> dict[str, str]:
...
-        """Remove hooks and reload all block weights from the checkpoint."""
-        self._remove_hooks()
-        self._load_all_blocks(model, block_names)
-        self._model_ref = None
+    def __init__(
+        self,
+        enabled: bool = True,
+        mode: str = "offload",
+        model_dir: Optional[str] = None,
+        prefix: str = "autoround_offload",
```
Comment (wenhuach21) [inline auto_round/utils/offload.py]: prefix is not very descriptive for developers. Would offload_dir_prefix be clearer, or names like that?

Comment (wenhuach21) [inline auto_round/utils/offload.py]: autoround_offload->ar_offload

File: test/test_cpu/core/test_low_cpu_mem_options.py [hunk 17]
```diff
@@ -64,7 +64,9 @@ def test_offloader_disabled_when_low_cpu_mem_false(self, tiny_opt_model_path):
         )
         assert autoround.low_cpu_mem_usage is False
         # offload with list should be a no-op when offloader is disabled
-        autoround._offloader.offload(autoround.model, [["model.layers.0"]], clear_memory=True, device_list=autoround.device_list)
+        autoround._offloader.offload(
```
Comment (wenhuach21) [inline test/test_cpu/core/test_low_cpu_mem_options.py:67]: better change str to list of str in the function rather than asking the user to provide a list

Comment (lvliang-intel) [reply test/test_cpu/core/test_low_cpu_mem_options.py:67]: ‘offload’ already accepts names Union[str, list[str], list[list[str]]], this is just for test.
